File: src/collector.py
"""
Orchestrate: list S3 log objects, download, parse, and export as OTLP logs to Grafana Alloy.
"""
import argparse
import os
import sys
from datetime import datetime, timezone
from typing import List, Optional

from botocore.exceptions import TokenRetrievalError

from .cdn_log_parser import parse_gcore_log_line
from .otel_exporter import create_logger_provider, emit_parsed_log
from .s3_source import list_log_objects, list_raw_keys, stream_log_lines, verify_bucket_access


def _normalize_prefix(p: str) -> str:
    return p if p.endswith("/") else (p + "/")


def _iter_keys(
    bucket: str,
    single_key: Optional[str],
    prefixes: List[str],
    since: Optional[datetime],
    aws_profile: Optional[str],
    max_objects: Optional[int],
):
    """Yield (key, last_modified, size). Either one key or from list_log_objects over prefixes."""
    if single_key:
        yield (single_key, datetime.now(timezone.utc), 0)
        return
    count = 0
    for prefix in prefixes:
        for key, last_mod, size in list_log_objects(
            bucket, prefix=prefix, since=since, profile_name=aws_profile
        ):
            yield key, last_mod, size
            count += 1
            if max_objects and count >= max_objects:
                return


def run(
    bucket: str,
    prefixes: List[str],
    since_minutes: Optional[int] = None,
    endpoint: Optional[str] = None,
    insecure: bool = True,
    service_name: str = "cdn-logs-collector",
    aws_profile: Optional[str] = None,
    dry_run: bool = False,
    max_objects: Optional[int] = None,
    max_lines_per_file: Optional[int] = None,
    verbose: bool = False,
    inspect_cmcd: bool = False,
    key: Optional[str] = None,
) -> int:
    """
    List S3 log objects under each prefix (or fetch a single key), parse each line as G-Core CDN log, emit as OTLP logs.
    Returns number of log records emitted (or that would be emitted if dry_run).
    """
    since = None
    if since_minutes is not None:
        from datetime import timedelta
        since = datetime.now(timezone.utc) - timedelta(minutes=since_minutes)

    prefixes = [_normalize_prefix(p) for p in prefixes]
    keys_iter = _iter_keys(bucket, key, prefixes, since, aws_profile, max_objects if not key else 1)

    if dry_run:
        count = 0
        for key_item, _last_mod, size in keys_iter:
            print(f"Would process: s3://{bucket}/{key_item} ({size} bytes)", file=sys.stderr)
            count += 1
        return 0

    if inspect_cmcd:
        # Only scan logs and print lines that have CMCD query strings (no OTLP export)
        cmcd_count = 0
        total_lines = 0
        objects_processed = 0
        if verbose and key:
            print(f"Fetching single key: s3://{bucket}/{key}", file=sys.stderr)
        for key_item, _last_mod, _size in keys_iter:
            if verbose and not key and objects_processed == 0:
                since_msg = f"since last {since_minutes} min" if since_minutes is not None else "all time"
                path = f"s3://{bucket}{prefixes[0]}" if prefixes and prefixes[0].startswith("/") else f"s3://{bucket}/{prefixes[0] if prefixes else ''}"
                print(f"Listing {path} ({since_msg})...", file=sys.stderr)
            line_count = 0
            for line in stream_log_lines(bucket, key_item, profile_name=aws_profile):
                if max_lines_per_file and line_count >= max_lines_per_file:
                    break
                parsed = parse_gcore_log_line(line)
                if parsed:
                    total_lines += 1
                    cmcd_attrs = {k: v for k, v in parsed.attributes.items() if k.startswith("cmcd.")}
                    if cmcd_attrs:
                        cmcd_count += 1
                        request = getattr(parsed, "request", "") or parsed.attributes.get("cdn.request", "")
                        print(f"request: {request}", file=sys.stderr)
                        print(f"CMCD:    {cmcd_attrs}", file=sys.stderr)
                        print("---", file=sys.stderr)
                line_count += 1
            objects_processed += 1
        print(
            f"Inspect summary: {cmcd_count} lines with CMCD out of {total_lines} total (from {objects_processed} objects)",
            file=sys.stderr,
        )
        return 0

    provider, logger = create_logger_provider(
        service_name=service_name,
        endpoint=endpoint or os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT"),
        insecure=insecure,
    )
    emitted = 0
    objects_processed = 0
    total_lines_read = 0
    try:
        for key_item, _last_mod, _size in keys_iter:
            if verbose:
                if key:
                    print(f"Processing: s3://{bucket}/{key_item}", file=sys.stderr)
                elif objects_processed == 0:
                    since_msg = f"since last {since_minutes} min" if since_minutes is not None else "all time"
                    path = f"s3://{bucket}{prefixes[0]}" if prefixes and prefixes[0].startswith("/") else f"s3://{bucket}/{prefixes[0] if prefixes else ''}"
                    print(f"Listing {path} ({since_msg})...", file=sys.stderr)
            line_count = 0
            parsed_count = 0
            for line in stream_log_lines(bucket, key_item, profile_name=aws_profile):
                if max_lines_per_file and line_count >= max_lines_per_file:
                    break
                parsed = parse_gcore_log_line(line)
                if parsed:
                    cmcd_attrs = {k: v for k, v in parsed.attributes.items() if k.startswith("cmcd.")}
                    if cmcd_attrs:
                        print(f"CMCD: {cmcd_attrs}", file=sys.stderr)
                    emit_parsed_log(logger, parsed, s3_key=key_item)
                    emitted += 1
                    parsed_count += 1
                line_count += 1
            total_lines_read += line_count
            if verbose:
                print(f"  {key_item}: {line_count} lines, {parsed_count} parsed", file=sys.stderr)
            objects_processed += 1
            if max_objects and objects_processed >= max_objects:
                break
    finally:
        provider.shutdown()
    if verbose:
        print(
            f"Summary: {objects_processed} objects, {total_lines_read} lines read, {emitted} log records emitted",
            file=sys.stderr,
        )
        if objects_processed == 0:
            print(
                "Debug: 0 objects matched. Listing raw keys (no suffix filter) to check prefix/bucket...",
                file=sys.stderr,
            )
            for p in prefixes:
                raw = list_raw_keys(bucket, prefix=p, max_keys=5, profile_name=aws_profile)
                if not raw and not p.startswith("/"):
                    # Some S3-compatible backends use keys with leading slash
                    raw = list_raw_keys(bucket, prefix="/" + p, max_keys=5, profile_name=aws_profile)
                if not raw:
                    print(f"  (no keys under prefix {p!r})", file=sys.stderr)
                    continue
                for key, last_mod, size in raw:
                    print(f"  key={key!r}  last_modified={last_mod!s}  size={size}", file=sys.stderr)
                if since is not None:
                    print(f"  Cutoff was: last_modified >= {since!s} (UTC)", file=sys.stderr)
    return emitted


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Transform CDN logs from S3 to OpenTelemetry and send to Grafana Alloy"
    )
    parser.add_argument(
        "--bucket",
        default=os.environ.get("CDN_LOGS_BUCKET", "amzn-gcore-logs"),
        help="S3 bucket name (default: CDN_LOGS_BUCKET or amzn-gcore-logs)",
    )
    parser.add_argument(
        "--prefix",
        action="append",
        default=None,
        dest="prefixes",
        metavar="PREFIX",
        help="S3 prefix for log objects; can be repeated. Default when not set: gcore/logs/ and 5gemerge/logs/",
    )
    parser.add_argument(
        "--since-minutes",
        type=int,
        default=None,
        help="Only process objects modified in the last N minutes",
    )
    parser.add_argument(
        "--endpoint",
        default=os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT", "localhost:4317"),
        help="OTLP gRPC endpoint (default: localhost:4317 for Alloy)",
    )
    parser.add_argument(
        "--no-insecure",
        action="store_true",
        help="Use TLS for OTLP (default is insecure=true)",
    )
    parser.add_argument(
        "--service-name",
        default=os.environ.get("OTEL_SERVICE_NAME", "cdn-logs-collector"),
        help="Service name for resource (default: cdn-logs-collector)",
    )
    parser.add_argument(
        "--aws-profile",
        default=os.environ.get("AWS_PROFILE"),
        help="AWS profile for S3 (default: AWS_PROFILE)",
    )
    parser.add_argument(
        "--region",
        default=os.environ.get("AWS_REGION", "eu-west-1"),
        help="AWS region for the bucket (default: eu-west-1)",
    )
    parser.add_argument(
        "--verify",
        action="store_true",
        help="Verify read access to the bucket and exit (no log processing)",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Only list S3 objects that would be processed",
    )
    parser.add_argument(
        "--max-objects",
        type=int,
        default=None,
        help="Max S3 objects to process (for testing)",
    )
    parser.add_argument(
        "--max-lines-per-file",
        type=int,
        default=None,
        help="Max lines to process per file (for testing)",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        action="store_true",
        help="Print progress: objects processed, lines read, parse counts",
    )
    parser.add_argument(
        "--inspect-cmcd",
        action="store_true",
        help="Only scan logs and print lines that have CMCD query strings (no OTLP export). Use to verify CMCD in requests.",
    )
    parser.add_argument(
        "--key",
        metavar="S3_KEY",
        default=None,
        help="Fetch and process only this S3 object key (e.g. /5gemerge/logs/2026/02/23/23/08/12/file_access.log.gz). Use with --inspect-cmcd to inspect one file.",
    )
    args = parser.parse_args()

    default_prefixes = ["/gcore/logs/", "/5gemerge/logs/"]
    if os.environ.get("CDN_LOGS_PREFIX"):
        default_prefixes = [os.environ.get("CDN_LOGS_PREFIX", "").rstrip("/") + "/"]
    prefixes = args.prefixes if args.prefixes is not None else default_prefixes

    try:
        if args.verify:
            ok = verify_bucket_access(
                bucket=args.bucket,
                region=args.region,
                profile_name=args.aws_profile,
                prefix=prefixes[0] if prefixes else "/5gemerge/logs/",
            )
            return 0 if ok else 1

        emitted = run(
            bucket=args.bucket,
            prefixes=prefixes,
            since_minutes=args.since_minutes,
            endpoint=args.endpoint,
            insecure=not args.no_insecure,
            service_name=args.service_name,
            aws_profile=args.aws_profile,
            dry_run=args.dry_run,
            max_objects=args.max_objects,
            max_lines_per_file=args.max_lines_per_file,
            verbose=args.verbose,
            inspect_cmcd=args.inspect_cmcd,
            key=args.key,
        )
        if not args.dry_run and not args.inspect_cmcd:
            print(f"Emitted {emitted} log records to {args.endpoint}", file=sys.stderr)
        return 0
    except TokenRetrievalError:
        print("AWS SSO token has expired. Log in again with:", file=sys.stderr)
        print("  aws sso login", file=sys.stderr)
        if args.aws_profile:
            print(f"  (or: aws sso login --profile {args.aws_profile})", file=sys.stderr)
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: ts/src/collector.ts
#!/usr/bin/env node
/**
 * Orchestrate: list S3 log objects, download, parse, and export as OTLP logs to Grafana Alloy.
 */

import { parseGcoreLogLine } from './cdn-log-parser.js';
import { createLoggerProvider, emitParsedLog } from './otel-exporter.js';
import {
  listLogObjects,
  listRawKeys,
  streamLogLines,
  verifyBucketAccess,
  type S3ObjectInfo,
} from './s3-source.js';

const DEFAULT_BUCKET = process.env.CDN_LOGS_BUCKET ?? 'amzn-gcore-logs';
const DEFAULT_PREFIXES = ['/gcore/logs/', '/5gemerge/logs/'];
const DEFAULT_ENDPOINT = process.env.OTEL_EXPORTER_OTLP_ENDPOINT ?? 'localhost:4317';

function normalizePrefix(p: string): string {
  return p.endsWith('/') ? p : p + '/';
}

async function* iterKeys(
  bucket: string,
  singleKey: string | null,
  prefixes: string[],
  since: Date | null,
  profile: string | undefined,
  maxObjects: number | null
): AsyncGenerator<S3ObjectInfo> {
  if (singleKey) {
    yield { key: singleKey, lastModified: new Date(), size: 0 };
    return;
  }
  let count = 0;
  for (const prefix of prefixes) {
    for await (const obj of listLogObjects(bucket, prefix, since ?? undefined, undefined, profile)) {
      yield obj;
      count += 1;
      if (maxObjects != null && count >= maxObjects) return;
    }
  }
}

export interface RunOptions {
  bucket: string;
  prefixes: string[];
  sinceMinutes: number | null;
  endpoint: string | null;
  insecure: boolean;
  serviceName: string;
  awsProfile: string | undefined;
  dryRun: boolean;
  maxObjects: number | null;
  maxLinesPerFile: number | null;
  verbose: boolean;
  inspectCmcd: boolean;
  key: string | null;
}

export interface CliArgs extends RunOptions {
  verify: boolean;
  region: string;
}

export async function run(options: RunOptions): Promise<number> {
  const {
    bucket,
    prefixes,
    sinceMinutes,
    endpoint,
    insecure,
    serviceName,
    awsProfile,
    dryRun,
    maxObjects,
    maxLinesPerFile,
    verbose,
    inspectCmcd,
    key,
  } = options;

  const since =
    sinceMinutes != null
      ? new Date(Date.now() - sinceMinutes * 60 * 1000)
      : null;
  const normalizedPrefixes = prefixes.map(normalizePrefix);
  const keysGen = iterKeys(
    bucket,
    key,
    normalizedPrefixes,
    since,
    awsProfile,
    key ? 1 : maxObjects
  );

  if (dryRun) {
    for await (const obj of keysGen) {
      console.error(`Would process: s3://${bucket}/${obj.key} (${obj.size} bytes)`);
    }
    return 0;
  }

  if (inspectCmcd) {
    let cmcdCount = 0;
    let totalLines = 0;
    let objectsProcessed = 0;
    if (verbose && key) {
      console.error(`Fetching single key: s3://${bucket}/${key}`);
    }
    for await (const obj of keysGen) {
      if (verbose && !key && objectsProcessed === 0) {
        const sinceMsg = sinceMinutes != null ? `since last ${sinceMinutes} min` : 'all time';
        const path = normalizedPrefixes[0]?.startsWith('/')
          ? `s3://${bucket}${normalizedPrefixes[0]}`
          : `s3://${bucket}/${normalizedPrefixes[0] ?? ''}`;
        console.error(`Listing ${path} (${sinceMsg})...`);
      }
      let lineCount = 0;
      for await (const line of streamLogLines(bucket, obj.key, awsProfile)) {
        if (maxLinesPerFile != null && lineCount >= maxLinesPerFile) break;
        const parsed = parseGcoreLogLine(line);
        if (parsed) {
          totalLines += 1;
          const cmcdAttrs = Object.fromEntries(
            Object.entries(parsed.attributes).filter(([k]) => k.startsWith('cmcd.'))
          );
          if (Object.keys(cmcdAttrs).length > 0) {
            cmcdCount += 1;
            const request = parsed.request || parsed.attributes['cdn.request'] || '';
            console.error(`request: ${request}`);
            console.error(`CMCD:    ${JSON.stringify(cmcdAttrs)}`);
            console.error('---');
          }
        }
        lineCount += 1;
      }
      objectsProcessed += 1;
    }
    console.error(
      `Inspect summary: ${cmcdCount} lines with CMCD out of ${totalLines} total (from ${objectsProcessed} objects)`
    );
    return 0;
  }

  const { provider, logger } = createLoggerProvider(
    serviceName,
    endpoint ?? undefined,
    insecure
  );
  let emitted = 0;
  let objectsProcessed = 0;
  let totalLinesRead = 0;

  try {
    for await (const obj of keysGen) {
      if (verbose) {
        if (key) {
          console.error(`Processing: s3://${bucket}/${obj.key}`);
        } else if (objectsProcessed === 0) {
          const sinceMsg = sinceMinutes != null ? `since last ${sinceMinutes} min` : 'all time';
          const path = normalizedPrefixes[0]?.startsWith('/')
            ? `s3://${bucket}${normalizedPrefixes[0]}`
            : `s3://${bucket}/${normalizedPrefixes[0] ?? ''}`;
          console.error(`Listing ${path} (${sinceMsg})...`);
        }
      }
      let lineCount = 0;
      let parsedCount = 0;
      for await (const line of streamLogLines(bucket, obj.key, awsProfile)) {
        if (maxLinesPerFile != null && lineCount >= maxLinesPerFile) break;
        const parsed = parseGcoreLogLine(line);
        if (parsed) {
          const cmcdAttrs = Object.fromEntries(
            Object.entries(parsed.attributes).filter(([k]) => k.startsWith('cmcd.'))
          );
          if (Object.keys(cmcdAttrs).length > 0 && verbose) {
            console.error(`CMCD: ${JSON.stringify(cmcdAttrs)}`);
          }
          emitParsedLog(logger, parsed, obj.key);
          emitted += 1;
          parsedCount += 1;
        }
        lineCount += 1;
      }
      totalLinesRead += lineCount;
      if (verbose) {
        console.error(`  ${obj.key}: ${lineCount} lines, ${parsedCount} parsed`);
      }
      objectsProcessed += 1;
      if (maxObjects != null && objectsProcessed >= maxObjects) break;
    }
  } finally {
    await provider.shutdown();
  }

  if (verbose) {
    console.error(
      `Summary: ${objectsProcessed} objects, ${totalLinesRead} lines read, ${emitted} log records emitted`
    );
    if (objectsProcessed === 0) {
      console.error('Debug: 0 objects matched. Listing raw keys (no suffix filter)...');
      for (const p of normalizedPrefixes) {
        const raw = await listRawKeys(bucket, p, 5, awsProfile);
        const alt = !p.startsWith('/') ? await listRawKeys(bucket, '/' + p, 5, awsProfile) : [];
        const list = raw.length ? raw : alt;
        if (!list.length) {
          console.error(`  (no keys under prefix ${JSON.stringify(p)})`);
          continue;
        }
        for (const o of list) {
          console.error(`  key=${JSON.stringify(o.key)}  last_modified=${o.lastModified.toISOString()}  size=${o.size}`);
        }
        if (since) console.error(`  Cutoff was: last_modified >= ${since.toISOString()} (UTC)`);
      }
    }
  }
  return emitted;
}

function parseArgs(): CliArgs {
  const argv = process.argv.slice(2);
  let verify = false;
  let dryRun = false;
  let noInsecure = false;
  let verbose = false;
  let inspectCmcd = false;
  let bucket = DEFAULT_BUCKET;
  let prefixes: string[] | null = null;
  let sinceMinutes: number | null = null;
  let endpoint: string | null = DEFAULT_ENDPOINT;
  let serviceName = process.env.OTEL_SERVICE_NAME ?? 'cdn-logs-collector';
  let awsProfile = process.env.AWS_PROFILE;
  let region = process.env.AWS_REGION ?? 'eu-west-1';
  let maxObjects: number | null = null;
  let maxLinesPerFile: number | null = null;
  let key: string | null = null;

  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === '--verify') verify = true;
    else if (arg === '--dry-run') dryRun = true;
    else if (arg === '--no-insecure') noInsecure = true;
    else if (arg === '-v' || arg === '--verbose') verbose = true;
    else if (arg === '--inspect-cmcd') inspectCmcd = true;
    else if (arg === '--bucket' && argv[i + 1]) {
      bucket = argv[++i];
    } else if ((arg === '--prefix' || arg === '-p') && argv[i + 1]) {
      if (!prefixes) prefixes = [];
      prefixes.push(argv[++i]);
    } else if (arg === '--since-minutes' && argv[i + 1]) {
      sinceMinutes = parseInt(argv[++i], 10);
    } else if (arg === '--endpoint' && argv[i + 1]) {
      endpoint = argv[++i];
    } else if (arg === '--service-name' && argv[i + 1]) {
      serviceName = argv[++i];
    } else if (arg === '--aws-profile' && argv[i + 1]) {
      awsProfile = argv[++i];
    } else if (arg === '--region' && argv[i + 1]) {
      region = argv[++i];
    } else if (arg === '--max-objects' && argv[i + 1]) {
      maxObjects = parseInt(argv[++i], 10);
    } else if (arg === '--max-lines-per-file' && argv[i + 1]) {
      maxLinesPerFile = parseInt(argv[++i], 10);
    } else if (arg === '--key' && argv[i + 1]) {
      key = argv[++i];
    }
  }

  if (process.env.CDN_LOGS_PREFIX) {
    prefixes = [process.env.CDN_LOGS_PREFIX.replace(/\/+$/, '') + '/'];
  } else if (!prefixes) {
    prefixes = [...DEFAULT_PREFIXES];
  }

  return {
    verify,
    bucket,
    prefixes,
    sinceMinutes,
    endpoint,
    insecure: !noInsecure,
    serviceName,
    awsProfile,
    region,
    dryRun,
    maxObjects,
    maxLinesPerFile,
    verbose,
    inspectCmcd,
    key,
  } as CliArgs;
}

async function main(): Promise<number> {
  const args = parseArgs();

  if (args.verify) {
    const ok = await verifyBucketAccess(
      args.bucket,
      args.region,
      args.awsProfile,
      args.prefixes[0] ?? '/5gemerge/logs/'
    );
    return ok ? 0 : 1;
  }

  try {
    const emitted = await run({
      bucket: args.bucket,
      prefixes: args.prefixes,
      sinceMinutes: args.sinceMinutes,
      endpoint: args.endpoint,
      insecure: args.insecure,
      serviceName: args.serviceName,
      awsProfile: args.awsProfile,
      dryRun: args.dryRun,
      maxObjects: args.maxObjects,
      maxLinesPerFile: args.maxLinesPerFile,
      verbose: args.verbose,
      inspectCmcd: args.inspectCmcd,
      key: args.key,
    });
    if (!args.dryRun && !args.inspectCmcd) {
      console.error(`Emitted ${emitted} log records to ${args.endpoint}`);
    }
    return 0;
  } catch (e: unknown) {
    const err = e as { name?: string; message?: string };
    if (err.name === 'TokenRetrievalError' || (err.message && err.message.includes('Token') && err.message.includes('expired'))) {
      console.error('AWS SSO token has expired. Log in again with:');
      console.error('  aws sso login');
      if (args.awsProfile) console.error(`  (or: aws sso login --profile ${args.awsProfile})`);
      return 1;
    }
    throw e;
  }
}

main()
  .then((code) => process.exit(code))
  .catch((err) => {
    console.error(err);
    process.exit(1);
  });
